File: packages/ccode/src/api/client/mcp.ts
/**
 * MCP API Client - Model Context Protocol via the zero-api service
 */

import z from "zod"
import { getClient, type ZeroClient } from "./index"

// ============================================================================
// MCP Types (matching zero-api)
// ============================================================================

/** MCP tool info */
export const McpToolInfo = z.object({
  name: z.string(),
  description: z.string(),
  input_schema: z.record(z.string(), z.unknown()),
})

export type McpToolInfo = z.infer<typeof McpToolInfo>

/** MCP content block */
export const McpContent = z.discriminatedUnion("type", [
  z.object({
    type: z.literal("text"),
    text: z.string(),
  }),
  z.object({
    type: z.literal("image"),
    data: z.string(),
    mime_type: z.string(),
  }),
  z.object({
    type: z.literal("resource"),
    uri: z.string(),
    text: z.string().optional(),
  }),
])

export type McpContent = z.infer<typeof McpContent>

/** MCP call request */
export const McpCallRequest = z.object({
  name: z.string(),
  arguments: z.record(z.string(), z.unknown()),
})

export type McpCallRequest = z.infer<typeof McpCallRequest>

/** MCP call response */
export const McpCallResponse = z.object({
  content: z.array(McpContent),
  is_error: z.boolean(),
})

export type McpCallResponse = z.infer<typeof McpCallResponse>

// ============================================================================
// MCP Client
// ============================================================================

/**
 * MCP API client
 */
export class McpClient {
  constructor(private client: ZeroClient = getClient()) {}

  /**
   * List available MCP tools
   */
  async listTools(): Promise<McpToolInfo[]> {
    return this.client.get<McpToolInfo[]>("/api/v1/mcp/tools")
  }

  /**
   * Call an MCP tool
   */
  async call(request: McpCallRequest): Promise<McpCallResponse> {
    return this.client.post<McpCallResponse>("/api/v1/mcp/call", request)
  }

  /**
   * Call an MCP tool with typed arguments
   */
  async callTool(name: string, args: Record<string, unknown>): Promise<McpCallResponse> {
    return this.call({ name, arguments: args })
  }

  /**
   * Call a tool and extract text content
   */
  async callForText(name: string, args: Record<string, unknown>): Promise<string> {
    const response = await this.callTool(name, args)
    if (response.is_error) {
      const errorText = response.content
        .filter((c): c is { type: "text"; text: string } => c.type === "text")
        .map((c) => c.text)
        .join("\n")
      throw new Error(`MCP tool error: ${errorText}`)
    }
    return response.content
      .filter((c): c is { type: "text"; text: string } => c.type === "text")
      .map((c) => c.text)
      .join("\n")
  }
}

/**
 * Singleton MCP client
 */
let mcpClient: McpClient | null = null

/**
 * Get the MCP client
 */
export function getMcpClient(): McpClient {
  if (!mcpClient) {
    mcpClient = new McpClient()
  }
  return mcpClient
}

/**
 * Namespace for MCP API
 */
export namespace ZeroMcp {
  export const listTools = () => getMcpClient().listTools()
  export const call = (request: McpCallRequest) => getMcpClient().call(request)
  export const callTool = (name: string, args: Record<string, unknown>) => getMcpClient().callTool(name, args)
  export const callForText = (name: string, args: Record<string, unknown>) => getMcpClient().callForText(name, args)
}

File: services/zero-cli/src/server/api/routes/mcp.rs
//! MCP (Model Context Protocol) API endpoints

use std::sync::Arc;

use axum::{
    extract::State,
    http::StatusCode,
    Json,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use super::super::state::AppState;

/// MCP tool definition
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpToolInfo {
    pub name: String,
    pub description: String,
    pub input_schema: Value,
}

/// MCP tool call request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpCallRequest {
    pub name: String,
    pub arguments: Value,
}

/// MCP tool call response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct McpCallResponse {
    pub content: Vec<McpContent>,
    pub is_error: bool,
}

/// MCP content block
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum McpContent {
    #[serde(rename = "text")]
    Text { text: String },
    #[serde(rename = "image")]
    Image { data: String, mime_type: String },
    #[serde(rename = "resource")]
    Resource { uri: String, text: Option<String> },
}

/// List available MCP tools
pub async fn list_tools(
    State(_state): State<Arc<AppState>>,
) -> Json<Vec<McpToolInfo>> {
    let tools = vec![
        McpToolInfo {
            name: "grep".to_string(),
            description: "Search file contents with regex patterns".to_string(),
            input_schema: serde_json::json!({
                "type": "object",
                "properties": {
                    "pattern": {
                        "type": "string",
                        "description": "Regex pattern to search for"
                    },
                    "path": {
                        "type": "string",
                        "description": "Directory to search in"
                    },
                    "glob": {
                        "type": "string",
                        "description": "Glob pattern to filter files"
                    }
                },
                "required": ["pattern"]
            }),
        },
        McpToolInfo {
            name: "read".to_string(),
            description: "Read file contents".to_string(),
            input_schema: serde_json::json!({
                "type": "object",
                "properties": {
                    "file_path": {
                        "type": "string",
                        "description": "Path to the file to read"
                    },
                    "offset": {
                        "type": "integer",
                        "description": "Line number to start reading from"
                    },
                    "limit": {
                        "type": "integer",
                        "description": "Number of lines to read"
                    }
                },
                "required": ["file_path"]
            }),
        },
        McpToolInfo {
            name: "write".to_string(),
            description: "Write content to a file".to_string(),
            input_schema: serde_json::json!({
                "type": "object",
                "properties": {
                    "file_path": {
                        "type": "string",
                        "description": "Path to write to"
                    },
                    "content": {
                        "type": "string",
                        "description": "Content to write"
                    }
                },
                "required": ["file_path", "content"]
            }),
        },
        McpToolInfo {
            name: "ls".to_string(),
            description: "List directory contents".to_string(),
            input_schema: serde_json::json!({
                "type": "object",
                "properties": {
                    "path": {
                        "type": "string",
                        "description": "Directory path to list"
                    }
                },
                "required": ["path"]
            }),
        },
        McpToolInfo {
            name: "webfetch".to_string(),
            description: "Fetch content from a URL".to_string(),
            input_schema: serde_json::json!({
                "type": "object",
                "properties": {
                    "url": {
                        "type": "string",
                        "description": "URL to fetch"
                    }
                },
                "required": ["url"]
            }),
        },
    ];

    Json(tools)
}

/// Call an MCP tool
pub async fn call_tool(
    State(state): State<Arc<AppState>>,
    Json(request): Json<McpCallRequest>,
) -> Result<Json<McpCallResponse>, (StatusCode, Json<McpCallResponse>)> {
    let make_error = |msg: String| -> (StatusCode, Json<McpCallResponse>) {
        (StatusCode::BAD_REQUEST, Json(McpCallResponse {
            content: vec![McpContent::Text { text: msg }],
            is_error: true,
        }))
    };

    let result = match request.name.as_str() {
        "grep" => {
            let options: zero_core::GrepOptions = serde_json::from_value(request.arguments)
                .map_err(|e| make_error(format!("Invalid arguments: {}", e)))?;

            let result = state.grep.search(&options).await
                .map_err(|e| make_error(format!("Grep failed: {}", e)))?;

            serde_json::to_string_pretty(&result).unwrap_or_default()
        }
        "read" => {
            let file_path = request.arguments.get("file_path")
                .and_then(|v| v.as_str())
                .ok_or_else(|| make_error("Missing file_path".to_string()))?;

            let result = state.reader.read(std::path::Path::new(file_path), None)
                .map_err(|e| make_error(format!("Read failed: {}", e)))?;

            result.content
        }
        "write" => {
            let file_path = request.arguments.get("file_path")
                .and_then(|v| v.as_str())
                .ok_or_else(|| make_error("Missing file_path".to_string()))?;
            let content = request.arguments.get("content")
                .and_then(|v| v.as_str())
                .ok_or_else(|| make_error("Missing content".to_string()))?;

            state.writer.write(std::path::Path::new(file_path), content, None)
                .map_err(|e| make_error(format!("Write failed: {}", e)))?;

            format!("Successfully wrote to {}", file_path)
        }
        "ls" => {
            let path = request.arguments.get("path")
                .and_then(|v| v.as_str())
                .unwrap_or(".");

            let result = state.ls.list(std::path::Path::new(path), None)
                .map_err(|e| make_error(format!("Ls failed: {}", e)))?;

            result.output
        }
        "webfetch" => {
            let url = request.arguments.get("url")
                .and_then(|v| v.as_str())
                .ok_or_else(|| make_error("Missing url".to_string()))?;

            let options = zero_core::WebFetchOptions {
                url: url.to_string(),
                ..Default::default()
            };

            let result = state.web_fetcher.fetch(&options).await
                .map_err(|e| make_error(format!("Fetch failed: {}", e)))?;

            result.content
        }
        _ => return Err((
            StatusCode::NOT_FOUND,
            Json(McpCallResponse {
                content: vec![McpContent::Text {
                    text: format!("Unknown tool: {}", request.name),
                }],
                is_error: true,
            }),
        )),
    };

    Ok(Json(McpCallResponse {
        content: vec![McpContent::Text { text: result }],
        is_error: false,
    }))
}
